File: ios/SpentyAI/SpentyAI/Core/Models/Report.swift
import Foundation

struct ReportSummary: Identifiable, Codable {
    var id: String { "\(startDate ?? "")-\(endDate ?? "")" }
    var totalIncome: Double?
    var totalExpense: Double?
    var totalTransfers: Double?
    var net: Double?
    var transactionCount: Int?
    var topCategories: [ReportCategory]?
    var startDate: String?
    var endDate: String?

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        totalIncome = try? c.decodeIfPresent(Double.self, forKey: .totalIncome)
        totalExpense = try? c.decodeIfPresent(Double.self, forKey: .totalExpense)
        totalTransfers = try? c.decodeIfPresent(Double.self, forKey: .totalTransfers)
        net = try? c.decodeIfPresent(Double.self, forKey: .net)
        transactionCount = try? c.decodeIfPresent(Int.self, forKey: .transactionCount)
        topCategories = try? c.decodeIfPresent([ReportCategory].self, forKey: .topCategories)
        startDate = try? c.decodeIfPresent(String.self, forKey: .startDate)
        endDate = try? c.decodeIfPresent(String.self, forKey: .endDate)
    }
}

struct IncomeExpenseResponse: Codable {
    var totalIncome: Double?
    var totalExpense: Double?
    var net: Double?
    var incomeByCategory: [ReportCategory]?
    var expenseByCategory: [ReportCategory]?
    var startDate: String?
    var endDate: String?

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        totalIncome = try? c.decodeIfPresent(Double.self, forKey: .totalIncome)
        totalExpense = try? c.decodeIfPresent(Double.self, forKey: .totalExpense)
        net = try? c.decodeIfPresent(Double.self, forKey: .net)
        incomeByCategory = try? c.decodeIfPresent([ReportCategory].self, forKey: .incomeByCategory)
        expenseByCategory = try? c.decodeIfPresent([ReportCategory].self, forKey: .expenseByCategory)
        startDate = try? c.decodeIfPresent(String.self, forKey: .startDate)
        endDate = try? c.decodeIfPresent(String.self, forKey: .endDate)
    }
}

struct ReportPeriod: Codable, Identifiable {
    var id: String { month ?? UUID().uuidString }
    var month: String?
    var income: Double?
    var expense: Double?
    var net: Double?
    var count: Int?
    var transfer: Double?

    /// Convenience alias so views can still use `period`
    var period: String? { month }
    /// Convenience alias so views can still use `transactionCount`
    var transactionCount: Int? { count }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        month = try? c.decodeIfPresent(String.self, forKey: .month)
        income = try? c.decodeIfPresent(Double.self, forKey: .income)
        expense = try? c.decodeIfPresent(Double.self, forKey: .expense)
        net = try? c.decodeIfPresent(Double.self, forKey: .net)
        count = try? c.decodeIfPresent(Int.self, forKey: .count)
        transfer = try? c.decodeIfPresent(Double.self, forKey: .transfer)
    }
}

struct ReportCategory: Codable, Identifiable {
    var id: String { categoryId ?? categoryName ?? UUID().uuidString }
    var categoryId: String?
    var categoryName: String?
    var total: Double?
    var income: Double?
    var expense: Double?
    var count: Int?
    var subcategories: [ReportSubcategory]?

    /// Convenience alias so views can still use `name`
    var name: String? { categoryName }
    /// Convenience alias so views can still use `amount`
    var amount: Double? { total }
    /// Convenience alias so views can still use `transactionCount`
    var transactionCount: Int? { count }
    /// Backend does not send percentage; views compute it from totalCategoryAmount
    var percentage: Double? { nil }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        categoryId = try? c.decodeIfPresent(String.self, forKey: .categoryId)
        categoryName = try? c.decodeIfPresent(String.self, forKey: .categoryName)
        total = try? c.decodeIfPresent(Double.self, forKey: .total)
        income = try? c.decodeIfPresent(Double.self, forKey: .income)
        expense = try? c.decodeIfPresent(Double.self, forKey: .expense)
        count = try? c.decodeIfPresent(Int.self, forKey: .count)
        subcategories = try? c.decodeIfPresent([ReportSubcategory].self, forKey: .subcategories)
    }
}

/// Subcategory breakdown returned by the backend inside each ReportCategory.
struct ReportSubcategory: Codable, Identifiable {
    var id: String { subcategoryId ?? subcategoryName ?? UUID().uuidString }
    var subcategoryId: String?
    var subcategoryName: String?
    var total: Double?
    var income: Double?
    var expense: Double?
    var count: Int?

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        subcategoryId = try? c.decodeIfPresent(String.self, forKey: .subcategoryId)
        subcategoryName = try? c.decodeIfPresent(String.self, forKey: .subcategoryName)
        total = try? c.decodeIfPresent(Double.self, forKey: .total)
        income = try? c.decodeIfPresent(Double.self, forKey: .income)
        expense = try? c.decodeIfPresent(Double.self, forKey: .expense)
        count = try? c.decodeIfPresent(Int.self, forKey: .count)
    }
}

File: android-native/app/src/main/java/com/spentyai/app/core/models/Report.kt
package com.spentyai.app.core.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class ReportSummary(
    @SerialName("totalIncome") val totalIncome: Double? = null,
    @SerialName("totalExpense") val totalExpense: Double? = null,
    @SerialName("totalTransfers") val totalTransfers: Double? = null,
    val net: Double? = null,
    @SerialName("transactionCount") val transactionCount: Int? = null,
    @SerialName("topCategories") val topCategories: List<ReportCategory>? = null,
    @SerialName("startDate") val startDate: String? = null,
    @SerialName("endDate") val endDate: String? = null
)

@Serializable
data class ReportPeriod(
    val month: String? = null,
    val income: Double? = null,
    val expense: Double? = null,
    val net: Double? = null,
    val count: Int? = null,
    val transfer: Double? = null
) {
    val period: String? get() = month
    val transactionCount: Int? get() = count
}

@Serializable
data class ReportCategory(
    @SerialName("categoryId") val categoryId: String? = null,
    @SerialName("categoryName") val categoryName: String? = null,
    val total: Double? = null,
    val income: Double? = null,
    val expense: Double? = null,
    val count: Int? = null,
    val subcategories: List<ReportSubcategory>? = null
) {
    val name: String? get() = categoryName
    val amount: Double? get() = total
    val transactionCount: Int? get() = count
    val percentage: Double? get() = null
}

@Serializable
data class ReportSubcategory(
    @SerialName("subcategoryId") val subcategoryId: String? = null,
    @SerialName("subcategoryName") val subcategoryName: String? = null,
    val total: Double? = null,
    val income: Double? = null,
    val expense: Double? = null,
    val count: Int? = null
)

@Serializable
data class IncomeExpenseResponse(
    @SerialName("totalIncome") val totalIncome: Double? = null,
    @SerialName("totalExpense") val totalExpense: Double? = null,
    val net: Double? = null,
    @SerialName("incomeByCategory") val incomeByCategory: List<ReportCategory>? = null,
    @SerialName("expenseByCategory") val expenseByCategory: List<ReportCategory>? = null,
    @SerialName("startDate") val startDate: String? = null,
    @SerialName("endDate") val endDate: String? = null
)

@Serializable
data class PeriodsResponse(
    val periods: List<ReportPeriod> = emptyList()
)

@Serializable
data class CategoriesReportResponse(
    val categories: List<ReportCategory> = emptyList()
)
